controllers/ingressnodefirewall_controller.go

```diff
@@ -46,57 +56,181 @@ type IngressNodeFirewallReconciler struct {
+		}
+		err = r.Create(ctx, &ingressNodeFirewallNodeState)
+		if err != nil {
+			log.Error(err, "Failed to create new IngressNodeFirewallNodeState",
+				"ingressNodeFirewallNodeState.Namespace", ingressNodeFirewallNodeState.Namespace, "ingressNodeFirewallNodeState.Name", ingressNodeFirewallNodeState.Name)
+			return ctrl.Result{}, err
 		}
+		log.Info("Created object",
+			"ingressNodeFirewallNodeState.Namespace", ingressNodeFirewallNodeState.Namespace, "ingressNodeFirewallNodeState.Name", ingressNodeFirewallNodeState.Name)
+	}
+
+	return ctrl.Result{}, nil
+}
+
+// triggerReconciliation triggers reconciliation for the first ingressnodefwv1alpha1.IngressNodeFirewall object that it can
+// find. Triggering reconciliation for a single object suffices because the IngressNodeFirewall reconciler will list and reconcile
+// all dependant resources with each reconciliation.
+// TODO: Limit reconciliation requests -> only reconcile if object.GetLabels() matches an IngressNodeFirewall with a matching label.
+func (r *IngressNodeFirewallReconciler) triggerReconciliation(object client.Object) []reconcile.Request {
+	ingressNodeFirewallList := ingressnodefwv1alpha1.IngressNodeFirewallList{}
+	listOpts := []client.ListOption{
+		client.InNamespace(r.Namespace),
+	}
+	if err := r.List(context.TODO(), &ingressNodeFirewallList, listOpts...); err != nil {
+		return []reconcile.Request{}
 	}
 
-	if err := c.IngressNodeFwAttach(*instance.Spec.Interfaces, isDelete); err != nil {
-		logger.Error(err, "Fail to attach ingress firewall prog %s", err)
-		return err
+	reconcileRequests := make([]reconcile.Request, len(ingressNodeFirewallList.Items))
+	for _, nodeState := range ingressNodeFirewallList.Items {
+		reconcileRequests = append(reconcileRequests, reconcile.Request{
+			NamespacedName: types.NamespacedName{
+				Name:      nodeState.Name,
+				Namespace: nodeState.Namespace,
+			},
+		})
+		// Abort after the first object.
+		break
 	}
-	return nil
+	return reconcileRequests
 }
 
-// SetupWithManager sets up the controller with the Manager.
+// SetupWithManager sets up the controller with the Manager. In addition to watching IngressNodeFirewall
+// this also watches objects of Kind Node.
 func (r *IngressNodeFirewallReconciler) SetupWithManager(mgr ctrl.Manager) error {
 	return ctrl.NewControllerManagedBy(mgr).
 		For(&ingressnodefwv1alpha1.IngressNodeFirewall{}).
+		Watches(
+			&source.Kind{Type: &v1.Node{}},
+			handler.EnqueueRequestsFromMapFunc(r.triggerReconciliation)).
 		Complete(r)
 }
+
+// buildNodeStateSpecs reads a list of *ingressnodefwv1alpha1.IngressNodeFirewallList and builds an appropriate mapping of
+// node => IngressNodeFirewallNodeStateSpec.
+// TODO: settle on a logic for the merges.
+func (r *IngressNodeFirewallReconciler) buildNodeStateSpecs(ctx context.Context, ingressNodeFirewallList *ingressnodefwv1alpha1.IngressNodeFirewallList) (map[string]*ingressnodefwv1alpha1.IngressNodeFirewallNodeStateSpec, error) {
```

Comment:
why is return map value a pointer? Doesnt seem to need it.

Reply:
done